modules/navier_stokes/src/kernels/INSMomentumSUPG.C

```diff
@@ -0,0 +1,312 @@
+  params.addCoupledVar("w", 0, "z-velocity"); // only required in 3D
+  params.addRequiredCoupledVar("p", "pressure");
+
+  // Required parameters
+  params.addRequiredParam<RealVectorValue>("gravity", "Direction of the gravity vector");
+  params.addRequiredParam<unsigned>(
+      "component",
+      "0,1,2 depending on if we are solving the x,y,z component of the momentum equation");
+
+  // Optional parameters
+  params.addParam<MaterialPropertyName>("mu_name", "mu", "The name of the dynamic viscosity");
+  params.addParam<MaterialPropertyName>("rho_name", "rho", "The name of the density");
+
+  return params;
+}
+
+INSMomentumSUPG::INSMomentumSUPG(const InputParameters & parameters)
+  : Kernel(parameters),
+
+    // Coupled variables
+    _u_vel(coupledValue("u")),
+    _v_vel(coupledValue("v")),
+    _w_vel(coupledValue("w")),
+    _p(coupledValue("p")),
+
+    // Gradients
+    _grad_u_vel(coupledGradient("u")),
+    _grad_v_vel(coupledGradient("v")),
+    _grad_w_vel(coupledGradient("w")),
+    _grad_p(coupledGradient("p")),
+
+    // Variable numberings
+    _u_vel_var_number(coupled("u")),
+    _v_vel_var_number(coupled("v")),
+    _w_vel_var_number(coupled("w")),
+    _p_var_number(coupled("p")),
+
+    // parameters
+    _gravity(getParam<RealVectorValue>("gravity")),
+    _component(getParam<unsigned>("component")),
+
+    // Material properties
+    _mu(getMaterialProperty<Real>("mu_name")),
+    _rho(getMaterialProperty<Real>("rho_name"))
+{
+}
+
+Real
+INSMomentumSUPG::computeQpResidual()
+{
+  RealVectorValue U(_u_vel[_qp], _v_vel[_qp], _w_vel[_qp]);
+  Real alpha;
+  if (_mu[_qp] < std::numeric_limits<double>::epsilon())
+    alpha = 1;
+  else if (U.norm() < std::numeric_limits<double>::epsilon())
+  {
+    alpha = 0;
+    return 0;
+  }
+  else
+  {
+    Real grid_Pe = _current_elem->hmax() * U.norm() / (2. * _mu[_qp]);
+    alpha = 1. / std::tanh(grid_Pe) - 1. / grid_Pe;
+  }
+  Real PG_test = alpha * _current_elem->hmax() / 2. * U / U.norm() * _grad_test[_i][_qp];
+
+  Real convective_part = _rho[_qp] * U * _grad_u[_qp] * PG_test;
+
+  // Note that integrating pressure by parts would imply applying a second
+  // derivative to the test function, which we are not equipped to do (?)
+  Real pressure_part = _grad_p[_qp](_component) * PG_test;
+
+  // From Computer Mechanics in Applied Mechanics and Engineering 32 (1982) pg. 212:
+  // For reasonable (linear) element shapes the contribution of PG_test * viscous
+  // term is small and can be neglected. This is not generally the case for higher
+  // order elements (and note that we commonly use quadratic shape functions for
+  // velocity). However, including this term would again imply applying a second
+  // derivative either to the test function or to the variable u
+  // Real viscous_part = computeQpResidualViscousPart();
+
+  Real body_force_part = -_rho[_qp] * _gravity(_component) * PG_test;
+
+  return convective_part + pressure_part + /*viscous_part +*/ body_force_part;
```

Comment:
As a longer term design goal, I think the "unstabilized" `INSMomentum` Kernel and the `INSMomentumSUPG` kernels should share a common base class which provides the strong (and weak) forms of the residual in a single location. That way, if new terms are added to the equation, they can be added in a single place rather than two different files. Code would also be able to be shared, since some parts of the strong and weak residual are the same (the parts that are not integrated by parts).